### PR TITLE
fix https://github.com/goccmack/gocc/issues/23

### PR DESCRIPTION
**Pull request by Chyroc**

### CODE REVIEW

**Comment by awalterschulze** [comment]
Nice :)

As I understand it an empty production rule, will now be the same as using the "empty" keyword, is that correct?

If so I then I think we should discuss whether we want this behavior or gocc to report an error.  Either way, I think I am happy, but what do the others think?

And also either way I think we need a test, to make sure we don't regress on this behavior later.

Really good edit, find the exact spot to fix this :)

**Comment by Chyroc** [comment]
Yes, if so, then the following two methods are equivalent

```bnf
B
	: empty 
	| "b"		<< "b", nil >>
	;

B
	: 
	| "b"		<< "b", nil >>
	;
```

**Comment by shivansh** [comment]
I couldn't find an empty production rule in the bnf described in #23 ; not sure what led to the production symbols being empty. Taking a look.

**Comment by Chyroc** [comment]
@shivansh see https://github.com/goccmack/gocc/issues/23#issuecomment-264673903
```
LExp
	: //"\" APat LExp2 "->" Exp
	| "let" Decls "in" Exp
```

**Comment by awalterschulze** [comment]
Also this one
```
Foo
	:
	| "b" 
;
```

**Comment by shivansh** [comment]
Ah, nice catch!

@awalterschulze I guess since we already have `empty` keyword for defining an empty production, maybe gocc should report an error in cases like above (which I guess will usually occur by mistake). What do you think?

**Comment by awalterschulze** [comment]
Exactly, this is the debate :)

@goccmack and @mewmew do you have an opinion as well?

**Comment by Chyroc** [comment]
It is also an acceptable solution to think that the lack of an "empty" is a bug, and the explicit syntax may be helpful.
Then I will modify the code to give a clear error message

**Comment by goccmack** [comment]
I think an empty alternative should only be indicated by using the keyword, empty. An alternative consisting of an empty string should give a parse error.

**Comment by mewmew** [comment]
I also think an error should be reported, as *empty* can be used to explicitly mark that a production rule is empty.

**Comment by awalterschulze** [comment]
Thanks everyone.
@Chyroc there you have it.  Everyone thinks an error would be better.  
Would you mind changing the logic to report an error?

I think we all agree that the code we have in master clearly needs a fix for this situation.

I see that the functions that you edited don't include an easy way to report an error, so maybe this check can be done earlier in a separate sweep over the grammar.
I think there is already such a sweep happening to find unused productions.

**Comment by Chyroc** [comment]
now , will exit with code 1 and message `maybe you missing the "empty" keyword in "B : \t<<  >>"`

File: internal/test/t1/t1.bnf [hunk 1]
```diff
@@ -1,10 +1,16 @@
...
 	;
 
 B
-	: empty 
+	: empty
 	| "b"		<< "b", nil >>
+	;
+
+C
+	:
```
Comment (awalterschulze) [inline internal/test/t1/t1.bnf]: I thought that gocc should give an error instead of generating the code for this example?
So it is probably easier to make a separate test.

**Comment by Chyroc** [comment]
@awalterschulze add test for missing "empty".

File: internal/test/t1/t2.bnf [hunk 2]
```diff
@@ -0,0 +1,10 @@
+!whitespace : ' ' ;
```
Comment (awalterschulze) [inline internal/test/t1/t2.bnf]: Why not put this in a test/t2 folder?

File: main.go [hunk 3]
```diff
@@ -83,6 +83,8 @@ func main() {
 		writeTerminals(gSymbols, cfg)
 	}
 
+	checkEmptyKeyword(g)
```
Comment (awalterschulze) [inline main.go]: Maybe this can be placed with the consistency check and return an error there?

https://github.com/goccmack/gocc/commit/101acee99d75357ecdc5acd209ee2cec8a638cc9

Then we have all the ast checks on one place

**Comment by Chyroc** [comment]
@awalterschulze done

File: internal/ast/grammar.go [hunk 4]
```diff
@@ -56,6 +56,9 @@ func consistent(g *Grammar) (err error) {
 		defs[tok.id] = true
 	}
 	for _, prod := range g.SyntaxPart.ProdList {
+		if len(prod.Body.Symbols) == 0 {
+			return fmt.Errorf("maybe you missing the \"empty\" keyword in %q", prod)
```
Comment (awalterschulze) [inline internal/ast/grammar.go]: This is a great error message, that actually gives you a tip of how to fix your mistake.

Sorry to be pedantic:
```
fmt.Errorf("empty production alternative: Maybe you are missing the \"empty\" keyword in %q", prod)
```

File: internal/test/t2/t2_test.go [hunk 5]
```diff
@@ -0,0 +1,24 @@
+package t1
```
Comment (awalterschulze) [inline internal/test/t2/t2_test.go]: package t2

**Comment by Chyroc** [comment]
@awalterschulze done

File: internal/test/t2/t2_test.go [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+	var buf = new(bytes.Buffer)
+	cmd := exec.Command("gocc", "t2.bnf")
+	cmd.Stdout = buf
+	err := cmd.Run()
+
+	e, ok := err.(*exec.ExitError)
+	if !ok || e.Success() {
+		t.Fatalf("gocc t2.bnf should return with exit error, but get %v.", err)
+	}
+	expectedStderr := `maybe you missing the "empty" keyword in "B : \t<<  >>"`
```
Comment (awalterschulze) [inline internal/test/t2/t2_test.go]: Hmmm, seems tests are passing, but this string is incorrect, are we sure this test is being run?

Comment (Chyroc) [reply internal/test/t2/t2_test.go]: no, test is correct, test check `!strings.Contains(actual, expectedStderr)` not `actual == expectedStderr`

Comment (Chyroc) [reply internal/test/t2/t2_test.go]: and add log `t.Logf("actual:\t%s\nexpectedStderr:\t%s\n", actual, expectedStderr)`, print:
```
	t2_test.go:22: actual:	Parse error: Error: $(0)  @ 10:2 maybe you missing the "empty" keyword in "B : \t<<  >>"
		
		
		expectedStderr:	maybe you missing the "empty" keyword in "B : \t<<  >>"
```

Comment (awalterschulze) [reply internal/test/t2/t2_test.go]: I thought actual was now supposed to be `Maybe you are missing the \"empty\" keyword`

**Comment by awalterschulze** [comment]
Now we just need one more maintainer to approve and merge :)
